### PR TITLE
Upgraded use of AWS Ruby SDK for S3

### PR DESCRIPTION
**Pull request by amancevice**
When `S3DownloadStrategy` is detected in `DownloadStrategyDetector`, an attempt is made to install & require the `aws-sdk-s3` gem.

The AWS S3 SDK is used to sign the URL and handed off to `curl_download`.

This PR is meant as a companion to https://github.com/Homebrew/brew/pull/3841.

ATTN @MikeMcQuaid, I couldn't figure out how to do this in `formula_installer.rb` because I don't think the `formula` object has access to the `@url` of the formula at that time. However, I think putting it in the `DownloadStrategyDetector` effectively front-runs the sandbox.

I'm a newbie when it comes to writing rspec tests so I've done a pretty barebones job at this. If I get some guidance I can expand the tests.

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/download_strategy.rb [hunk 1]
```diff
@@ -1148,6 +1142,9 @@ def self.detect_from_url(url)
       SubversionDownloadStrategy
     when %r{^https?://(.+?\.)?sourceforge\.net/hgweb/}
       MercurialDownloadStrategy
+    when %r{^s3://}
+      require_aws_sdk
```
Comment (MikeMcQuaid) [inline Library/Homebrew/download_strategy.rb]: Why is this needed in two places?

File: Library/Homebrew/download_strategy.rb [hunk 2]
```diff
@@ -1169,4 +1166,11 @@ def self.detect_from_symbol(symbol)
       raise "Unknown download strategy #{symbol} was requested."
     end
   end
+
+  def self.require_aws_sdk
+    require "aws-sdk-s3"
+  rescue LoadError
```
Comment (MikeMcQuaid) [inline Library/Homebrew/download_strategy.rb]: Make this unconditionally run `install_gem_setup_path` instead of the `rescue`

File: Library/Homebrew/download_strategy.rb [hunk 3]
```diff
@@ -1169,4 +1166,11 @@ def self.detect_from_symbol(symbol)
...
+    Homebrew.install_gem_setup_path! "aws-sdk-s3", "~> 1.8", "ruby"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/download_strategy.rb]: Don't pass `ruby` here.

File: Library/Homebrew/download_strategy.rb [hunk 4]
```diff
@@ -473,17 +473,8 @@ def stage
...
-    begin
-      require "aws-sdk-v1"
-    rescue LoadError
-      onoe "Install the aws-sdk gem into the gem repo used by brew."
-      raise
-    end
-
-    if @url !~ %r{^https?://([^.].*)\.s3\.amazonaws\.com/(.+)$}
+    unless @url =~ %r{^https?://([^.].*)\.s3\.amazonaws\.com/(.+)$} ||
+           @url =~ %r{^s3://([^.].*?)/(.+)$}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/download_strategy.rb]: Could you keep the changes like this into a separate PR? It makes it easier to review and test this one. Thanks!

File: Library/Homebrew/download_strategy.rb (hunk 1 above)
Comment (amancevice) [reply Library/Homebrew/download_strategy.rb]: This is necessary to support both cases when the formula explicitly states `:using => S3DownloadStrategy` and again when it’s implied by using the `s3://` scheme in the download URL.

Ex:

```
# Traditional URL with :using
url “https://but.s3.anazonaws.com/key” :using => S3DownloadStrategy

# S3 URL
url “s3://bkt/key”
```

File: Library/Homebrew/download_strategy.rb (hunk 4 above)
Comment (amancevice) [reply Library/Homebrew/download_strategy.rb]: Sure

File: Library/Homebrew/download_strategy.rb (hunk 1 above)
Comment (amancevice) [reply Library/Homebrew/download_strategy.rb]: As noted above, I can separate these into different PRs, but I think it’s needed in both places

Comment (amancevice) [reply Library/Homebrew/download_strategy.rb]: As noted above, I can separate these into different PRs, but I think it’s needed in both places

File: Library/Homebrew/download_strategy.rb (hunk 3 above)
Comment (amancevice) [reply Library/Homebrew/download_strategy.rb]: Completely leave off the argument? I tried passsing `nil` and it threw an error at me.

File: Library/Homebrew/download_strategy.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/download_strategy.rb]: It would be good to have:

1. the gem changes to the existing code
2. the `s3://` extensions

be separate PRs, yeh.

File: Library/Homebrew/download_strategy.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/download_strategy.rb]: Try to omit the argument entirely and if that fails post both that and the `nil` errors here and we'll talk about an API fix. Thanks!

Comment (amancevice) [reply Library/Homebrew/download_strategy.rb]: After uninstalling the gem from the Homebrew bundle...

With `Homebrew.install_gem_setup_path! "aws-sdk-s3", "~> 1.8"` I get:

```
==> Installing or updating 'aws-sdk-s3' gem
Fetching: aws-sdk-s3-1.8.2.gem (100%)
Successfully installed aws-sdk-s3-1.8.2
1 gem installed
Error: The 'aws-sdk-s3' gem is installed but couldn't find 'aws-sdk-s3' in the PATH:
```

With `Homebrew.install_gem_setup_path! "aws-sdk-s3", "~> 1.8", nil` I get:

```
==> Installing or updating 'aws-sdk-s3' gem
Fetching: aws-sdk-s3-1.8.2.gem (100%)
Successfully installed aws-sdk-s3-1.8.2
1 gem installed
Error: no implicit conversion of nil into String
```

Comment (amancevice) [reply Library/Homebrew/download_strategy.rb]: Any advice, @MikeMcQuaid?

**Comment by MikeMcQuaid** [comment]
Looks good, nice work again @amancevice!
